crates/wit-component/fuzz/fuzz_targets/roundtrip-wit.rs

```diff
@@ -0,0 +1,502 @@
+#![no_main]
+
+use libfuzzer_sys::fuzz_target;
+use std::path::Path;
+use wit_component::*;
+
+fuzz_target!(|data: &[u8]| {
+    drop(env_logger::try_init());
+
+    let mut u = arbitrary::Unstructured::new(data);
+    let doc = match generate::document(&mut u) {
+        Ok(doc) => doc,
+        Err(_) => return,
+    };
+    let text = DocumentPrinter::default().print(&doc).unwrap();
+    write_file("doc.wit", &text);
+    let world = match doc.default_world() {
+        Ok(world) => world,
+        Err(_) => return,
+    };
+
+    let types_only_binary = ComponentEncoder::default()
+        .validate(true)
+        .types_only(true)
+        .document(doc.clone(), StringEncoding::UTF8)
+        .unwrap()
+        .encode()
+        .unwrap();
+    write_file("doc.types-only.wasm", &types_only_binary);
+    let (types_only_doc, _) = decode_world(&doc.worlds[world].name, &types_only_binary).unwrap();
+    let text_from_types_only = DocumentPrinter::default().print(&types_only_doc).unwrap();
+    write_file("doc.types-only.wit", &text_from_types_only);
+
+    let dummy = test_helpers::dummy_module(&doc);
+    write_file("doc.dummy.wasm", &dummy);
+    let normal_binary = ComponentEncoder::default()
+        .validate(true)
+        .module(&dummy)
+        .unwrap()
+        .document(doc.clone(), StringEncoding::UTF8)
+        .unwrap()
+        .encode()
+        .unwrap();
+    write_file("doc.normal.wasm", &normal_binary);
+    let (normal_doc, _) = decode_world(&doc.worlds[world].name, &normal_binary).unwrap();
+    let text_from_normal = DocumentPrinter::default().print(&normal_doc).unwrap();
+    write_file("doc.normal.wit", &text_from_normal);
+
+    if text_from_normal != text_from_types_only {
+        panic!("text not equal");
+    }
+});
+
+fn write_file(path: &str, contents: impl AsRef<[u8]>) {
+    if !log::log_enabled!(log::Level::Debug) {
+        return;
+    }
+    log::debug!("writing file {path}");
+    let contents = contents.as_ref();
+    let path = Path::new(path);
+    std::fs::write(path, contents).unwrap();
+    if path.extension().and_then(|s| s.to_str()) == Some("wasm") {
+        let path = path.with_extension("wat");
+        log::debug!("writing file {}", path.display());
+        std::fs::write(path, wasmprinter::print_bytes(&contents).unwrap()).unwrap();
+    }
+}
+
+mod generate {
```

Comment:
While theoretically any component should be able to get a `*.wit` document extracted from it I think we still have aways to go for that on the specification. For example named types aren't implemented in such a way to allow that to infallibly happen. Additionally the compnent generation hasn't been updated in awhile so it may not produce the most interesting components to extract a document from.

You're right though in that I would rather generate the text rather than an AST directly here. For now this is just an intial state of a fuzzer though since I didn't have the text parsing on-hand. I hope to improve this and perhaps merge it into the top-level fuzz targets eventually if it's useful enough. For now though it's only really useful for the development of this PR so far (although will be helpful in future refactorings)